src-input/duk_bi_buffer.c

```diff
@@ -114,6 +114,42 @@ static duk_uint16_t duk__buffer_elemtype_copy_compatible[9] = {
 };
 #endif  /* !DUK_USE_PREFER_SIZE */
 
+#if defined(DUK_USE_BUFFEROBJECT_SUPPORT)
+/* Buffer supported encodings */
+
+#define DUK_BUF_ENC_UNKNOWN              0
+#define DUK_BUF_ENC_UTF8                 1
+
+/* longest encoding string + 1 -- should be updated when longer strings are added */
+#define DUK_BUFFER_ENCODING_MAX_LEN      7
+
+#define DUK_BUFFER_ENCODING_COUNT        2
+DUK_LOCAL const char * const duk__buffer_encoding_names[DUK_BUFFER_ENCODING_COUNT] = {
+	"utf8",
+	"utf-8"
+};
+
+DUK_LOCAL const duk_int_t duk__buffer_encoding_type_from_name[DUK_BUFFER_ENCODING_COUNT] = {
+	DUK_BUF_ENC_UTF8,
+	DUK_BUF_ENC_UTF8
+};
+
+DUK_LOCAL duk_int_t duk__parse_string_encoding(const char *encoding) {
+	duk_uint8_t i;
+	char buf[DUK_BUFFER_ENCODING_MAX_LEN];
+	/* the valid nodejs buffer encodings only contain letters numbers and hyphens */
+	for (i = 0; i < DUK_BUFFER_ENCODING_MAX_LEN; ++i) {
+		if (encoding[i] == 0) { buf[i] = 0; break; }
+		buf[i] = (char) (encoding[i] | 0x20);
+	}
```

Comment:
This loop works incorrectly if `i` reaches the maximum encoding length without finding a NUL terminator in the input: `buf` will remain unterminated and it's used for a `strcmp()` later.

Also inputs like `utf8\u0000bar` are (technically) handled incorrectly, as the loop will terminate on the NUL.

Reply:
I wonder if it'd produce less code to just short circuit compare the input, e.g.:
```
if ((p[0] | 0x20) == 'u' && (p[1] | 0x20) == 't' && (p[2] | 0x20) == 'f') {
    if (p[3] == '8' && encoding_len == 4) {
        /* utf8 */
    } else if (p[3] == '-' && p[4] == '8' && encoding_len == 5) {
        /* utf8 */
    }
}
/* invalid */
```